pkg/reduce/data_forward.go

```diff
@@ -348,9 +348,15 @@ func (df *DataForward) Process(ctx context.Context, messages []*isb.ReadMessage)
 	if len(successfullyWrittenMessages) == 0 {
 		return
 	}
+
 	// ack successful messages
 	df.ackMessages(ctx, successfullyWrittenMessages)
 
+	// non-ack the failed messages, so that they will be retried in the next iteration
+	// if we don't do this, the failed messages will be retried after the ackWait time
+	// which will cause correctness issues
```

Comment:
```suggestion
	// no-ack the failed messages, so that they will be retried in the next iteration
	// if we don't do this, the failed messages will be retried after the ackWait time
	// which will cause correctness issues. We want these messages to be immediately retried.
	// When a message is retried, the offset remains the same, so an old message might jump out of the offset-timeline and cause the watermark to be -1.
	// The correctness is violated because the queue offset and message order are no longer monotonically increasing for those failed messages.
```